### PR TITLE
Application API Client Support for Config with --generation Option

### PR DESCRIPTION
**Pull request by manadart**
## Description of change

This patch threads the active generation through API clients from application commands.

- Commands for _deploy_ (including bundles), _bundle-diff_ and _set-plan_ always assume the current generation.
- Commands for application and charm config allow the user to add a _--generation_ option, which if not set will use the active generation from from local storage (defaulting to "current" here).
- Resources are not dealt with by the patch but will be generational in future.

Patches will follow this one to:
- Retrieve the active generation from local storage (depends on #9661).
- Ensure all usages of generation are passed in calls to the API server.

## QA steps

At this time, we want to ensure that all client operations involving generational settings are operational, but not backward-breaking. So we test a new client against an old controller.

- Bootstrap a controller with a version prior to this patch - develop HEAD will do fine.
- Build and install Juju from this patch.
- `export JUJU_DEV_FEATURE_FLAGS=generations`
- Deploy an old version of an application, for example `juju deploy cs:trusty/mariadb-6`
- Check that the following commands work correctly with and without the _--generation_ option:
  - `juju config mariadb` (plus get, set and unset).
  - `juju upgrade-charm` (should cause an upgrade in this example to cs:trusty/mariadb-7).
  - `juju export-bundle > temp.yaml` runs without error.
  - `juju diff-bundle temp.yaml` runs without error.
  - `juju set-plan mariadb www.google.com` will return an error due to the URL, but will run correctly.

## Documentation changes

To come with _Generations_ feature generally.

## Bug reference

N/A

### CODE REVIEW

File: api/application/client.go [hunk 1]
```diff
@@ -488,10 +489,10 @@ func (c *Client) DestroyUnits(in DestroyUnitsParams) ([]params.DestroyUnitResult
 // TODO(axw) 2017-03-16 #1673323
 // Drop this in Juju 3.0.
 func (c *Client) DestroyDeprecated(application string) error {
-	params := params.ApplicationDestroy{
+	args := params.ApplicationDestroy{
```
Comment (manadart) [inline api/application/client.go]: Drive-by import collision fixes; sorry for the noise.

**Comment by manadart** [comment]
[Add generation argument to application config APIs](https://trello.com/c/mlFdCa0Q/987-add-generation-argument-to-application-config-apis)

**Comment by hmlanigan** [comment]
During QA, I bootstrapped 2.5.0 and ran config with --generation next.  It accepted the flag with no issues.  Shouldn't we have a message to say that the flag isn't supported with this controller version?

**Comment by manadart** [comment]
$$merge$$
